### ISSUE TITLE
Implementing a fresh node with the Master branch displays a root hash and bad block errors

### ISSUE THREAD

**Issue by bcamacho** [celo-org/celo-blockchain#1578]
# Issue

## Implementing a fresh node and synchronizing the node to the head of the chain.

Sync to head with a fresh node with the master branch displays a root hash and bad block errors. The Master branch is responding differently compared to the recent release of Celo-blockchain-1.3.2 with a fresh node on the test network. 

#### Tools used

- Geth
- CeloCLI

## Master branch

- Pulled Master branch to the local environment
- Tested with Windows, Linux, and OS X
- Compiled Geth using Make.
- Retrieve Geth version
- Configure command line parameters (flags)
- Synchronize node with the network
- Tested sync and account balance with CeloCLI

### Geth Version

__Command:__

```
./celo-blockchain/build/bin/geth version 
```

__Resonse:__

```
Celo
Version: 1.0.0-unstable
Git Commit: ab78658ae3a738d91d877f289e37fead2a0f1cf7
Git Commit Date: 20210527
Architecture: amd64
Protocol Versions: [66 65 64]
Go Version: go1.16.4
Operating System: darwin
GOPATH=
GOROOT=/usr/local/go
```

### Synchronizing node with the network

__Command:__

_*Note, coresdomain with * does not follow best practices. I’m using it to prevent cors issues for testing._

```
./celo-blockchain/build/bin/geth --alfajores --rpc --rpcapi eth,net,web3,debug,admin,personal --rpccorsdomain '*' --graphql --light.serve 90 --light.maxpeers 1000 --maxpeers 1100 --etherbase 0x0dD6D5c3cd2a3336f088981c7573eecEaF27E744 console 
```

### Terminal Response regarding Geth flags, rpc to http.

Forked repo, updated the readme to include new flags:  (https://github.com/celo-org/celo-blockchain/pull/1561)

__Resonse:__

```
1100 --etherbase 0x0dD6D5c3cd2a3336f088981c7573eecEaF27E744 console 
INFO [06-09|11:53:29.210] Starting Geth on Alfajores testnet... 
INFO [06-09|11:53:29.211] Maximum peer count                       ETH=100 LES=1000 total=1100
WARN [06-09|11:53:29.211] The flag --rpc is deprecated and will be removed in the future, please use --http 
WARN [06-09|11:53:29.211] The flag --rpccorsdomain is deprecated and will be removed in the future, please use --http.corsdomain 
WARN [06-09|11:53:29.211] The flag --rpcapi is deprecated and will be removed in the future, please use --http.api 
```

### Modified terminal command with adjusted parameters with http flags

__Command:__

```
./celo-blockchain/build/bin/geth --alfajores --http --http.api eth,net,web3,debug,admin,personal --http.corsdomain '*' --graphql --light.serve 90 --light.maxpeers 1000 --maxpeers 1100 --etherbase 0x0dD6D5c3cd2a3336f088981c7573eecEaF27E744 console
```

__Resonse:__

```
Error: invalid receipt root hash (remote: a9a8dd048b414dabf4bc9d1a7b3c412312aea98eb3312a1867c137355e28e8ea local: dd66bbaf74d66d73f06cb493d87a5e5e344e52a486abb7ec6bab62b5234762b0)
##############################
 
WARN [06-01|13:30:05.850] Synchronisation failed, dropping peer    peer=df431a8d3e701951 err="retrieved hash chain is invalid: invalid receipt root hash (remote: a9a8dd048b414dabf4bc9d1a7b3c412312aea98eb3312a1867c137355e28e8ea local: dd66bbaf74d66d73f06cb493d87a5e5e344e52a486abb7ec6bab62b5234762b0)"
INFO [06-01|13:30:05.903] Ethereum peer connected                  id=edfd5eba1691cfac conn=dyndial name=celo/v1.3.2-stable-91993404/linux-amd64/go1.16.4
WARN [06-01|13:30:05.926] Synchronisation failed, dropping peer    peer=86344b69c6d62286 err="action from bad peer ignored"
INFO [06-01|13:30:09.830] Ethereum peer connected                  id=7467fb552aa463e4 conn=inbound name=celo/v1.3.2-stable-91993404/linux-amd64/go1.14.12
ERROR[06-01|13:30:09.993] 
########## BAD BLOCK #########
Chain config: {ChainID: 44787 Homestead: 0 DAO: <nil> DAOSupport: true EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0 Petersburg: 0 Istanbul: 0 Churrito: 4960000, Donut: 4960000, Engine: istanbul}

Number: 5395481
Hash: 0x34d75059f65a6a9dec87bbf31af5d6daefb22f331e5813e37d0b943fd3b50351
	 0: cumulative: 199037 gas: 199037 contract: 0x0000000000000000000000000000000000000000 status: 1 tx: 0xd481ad963beb0b4f82c2999a95787fb1671f44cf369c0d540942770aba7a9dbc logs: [0xc006670b00 0xc006670c60] bloom: 00800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000100000000000204000000000000000000000000000000000000000000000000008000000000000000000000001000000000000000000000000040400000000000000000000000000000000000000000000000000000800000004000000000000000000002000000000000000000000000000000000004000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000 state: 
	 1: cumulative: 397172 gas: 198135 contract: 0x0000000000000000000000000000000000000000 status: 1 tx: 0x4ba426c5c18c79ebdf1d86021f99cef06936c1a061f7bc7280deb8fefd483176 logs: [0xc006670dc0 0xc006670e70] bloom: 00800000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000802000000000000000000000000800000040000000000000000000000000100000000000004000000000080000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000400008000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000 state: 
	 2: cumulative: 504410 gas: 107238 contract: 0x0000000000000000000000000000000000000000 status: 1 tx: 0x66567727a711a43eddcb7fc9269ed534ce3892a1e83e13b983a01d7d37bebb5f logs: [0xc006670f20 0xc006670fd0 0xc006671080 0xc0066711e0] bloom: 00000000000020000000000000000000000000000200000008000000000000000000000000004000000000000000000000000000000000000000000000100000000000000000000000000008000000000000000000000000000008000000000000400000000000080000000000008000000000000000000000000010000000000000008800000000000000000000000000400000000000040000000400000000000000000000000000000000000000000000000000000000000000000000000000000002000000000000000000000010000000000000000000000000000002000000000000000000000000000000000000000000000000000000008000000004 state: 
	 3: cumulative: 0 gas: 0 contract: 0x0000000000000000000000000000000000000000 status: 1 tx: 0x0000000000000000000000000000000000000000000000000000000000000000 logs: [0xc006671290] bloom: 00000000000200000000000000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 state: 


Error: invalid receipt root hash (remote: a9a8dd048b414dabf4bc9d1a7b3c412312aea98eb3312a1867c137355e28e8ea local: dd66bbaf74d66d73f06cb493d87a5e5e344e52a486abb7ec6bab62b5234762b0)
##############################
 
WARN [06-01|13:30:10.006] Synchronisation failed, dropping peer    peer=84c3c05abe51ab41 err="retrieved hash chain is invalid: invalid receipt root hash (remote: a9a8dd048b414dabf4bc9d1a7b3c412312aea98eb3312a1867c137355e28e8ea local: dd66bbaf74d66d73f06cb493d87a5e5e344e52a486abb7ec6bab62b5234762b0)"
INFO [06-01|13:30:10.089] Ethereum peer connected                  id=4fdbb859f2ed2638 conn=inbound name=celo/v1.1.0-stable/linux-amd64/go1.13.15
INFO [06-01|13:30:10.090] Ethereum handshake failed                id=4fdbb859f2ed2638 conn=inbound err="Fork ID rejected - remote needs update"
WARN [06-01|13:30:10.248] Synchronisation failed, dropping peer    peer=7467fb552aa463e4 err="action from bad peer ignored"
INFO [06-01|13:30:11.360] Deep froze chain segment                 blocks=4792 elapsed=757.811ms number=5305480 hash="a9eafd…843254"
INFO [06-01|13:30:13.104] Ethereum peer connected                  id=84c3c05abe51ab41 conn=dyndial name=celo/v1.3.2-stable/linux-amd64/go1.16.4
INFO [06-01|13:30:14.324] Ethereum peer connected                  id=86344b69c6d62286 conn=inbound name=celo/v1.3.2-stable-91993404/linux-amd64/go1.16.4
ERROR[06-01|13:30:14.864] 
########## BAD BLOCK #########
```

### Testing sync status with CeloCLI

__Command:__

Check if the node is synced

```
celocli node:synced
```

__Response:__

```
false
```

__Command:__

Check account balance

```
celocli account:balance 0x0dD6D5c3cd2a3336f088981c7573eecEaF27E744
```

__Response:__

```
 Error: Node is not currently synced. Run node:synced to check its status.
```

_Makes sense! Nodes not synced. Attempted to leave node running for 24 hours and monitored the data directory. After 24 hours, the node lacked progress which signaled to test with our recent official release._

## Celo blockchain v1.3.2-stable
[Downloaded official release of Celo-blockchain-1.3.2-stable](https://github.com/celo-org/celo-blockchain/releases)

- Decompressed archived to local environment
- Tested with Windows, Linux, and OS X
- Compiled Geth using Make.
- Retrieve Geth version
- Configure command line parameters (flags)
- Synchronize node with network
- Tested sync and account balance with CeloCLI

### Geth Version

__Command:__

```
./celo-blockchain/build/bin/geth version 
```

__Response:__

```
Celo
Version: 1.3.2-stable
Architecture: amd64
Protocol Versions: [66 65 64]
Go Version: go1.16.4
Operating System: darwin
GOPATH=
GOROOT=/usr/local/go

```

### Synchronizing node with network

__Command:__

_*Note, coresdomain with * does not follow best practices. I’m using to prevent cors issues for testing._

```
./celo-blockchain-1.3.2/build/bin/geth --alfajores --rpc --rpcapi eth,net,web3,debug,admin,personal --rpccorsdomain '*' --graphql --light.serve 90 --light.maxpeers 1000 --maxpeers 1100 --etherbase 0x0dD6D5c3cd2a3336f088981c7573eecEaF27E744 console 

```

__Response:__
Node synced, data directory incrementing as expected. Let the node sync for a few hours and tested with CeloCLI to check status. The result was a success, node synced, and console log look good 😊.
```
INFO [06-09|12:32:08.159] Imported new chain segment               blocks=1  txs=0 mgas=0.000 elapsed=14.335ms  mgasps=0.000  number=5584731 hash=e970e3…674846 dirty=2.53MiB
INFO [06-09|12:32:13.044] Imported new chain segment               blocks=1  txs=0 mgas=0.000 elapsed=13.663ms  mgasps=0.000  number=5584732 hash=3c525a…6cdea3 dirty=2.53MiB
INFO [06-09|12:32:18.029] Imported new chain segment               blocks=1  txs=0 mgas=0.000 elapsed=11.940ms  mgasps=0.000  number=5584733 hash=c7e6da…bd28b8 dirty=2.51MiB
```

### Testing sync status with CeloCLI

__Command:__

Check if node is synced
```
celocli node:synced
```

__Response:__
```
true
```

__Command:__

Check account balance
```
celocli account:balance 0x0dD6D5c3cd2a3336f088981c7573eecEaF27E744
```

__Response:__
```
All balances expressed in units of 10^-18.
lockedCELO: 0
pending: 0
CELO: 5000000000000000000
cUSD: 10000000000000000000
cEUR: 10000000000000000000
```

While posting this issue I did a quick test. The difference between my test now compared to the start of developing a fresh node is that my data directory has been populated to the head of the chain. I switched to the compiled master branch and tested it. The node responded as synced and the account balance was retrieved. Furthermore, the log output appeared normal (only tested for a quick minute as I populate this issue).

## Conclussion:

There might be an issue with our Master Branch when configuring a full node on the network. To resolve, utilize the latest release to sync to the head of the chain. In addition, parameter flags need to be modified between the two sources (Master Branch versus Release v1.3.2).

**Comment by piersy** [comment]
Hi @bcamacho, thanks for raising this, it looks like a problem has been introduced between the donut release and the tip of master. We are currently running some tests to try and pinpoint where exactly that occurred.

**Comment by bcamacho** [comment]
Wonderful, thanks for looking into this!

When you have an update, let me know and I will test on my end. 

The two tests I like to run are (after building geth):

1.  Sync node with fresh blockchain data
2. Sync node with snapshot data

**Comment by oneeman** [comment]
@bcamacho Thanks again for reporting this!  We've tracked down the cause of this and have a fix for it.  It turns out the problem goes away in later commits on master, but not because the original cause was fixed, but rather because the problem was just no longer hit due to other changes.  I've created a branch `oneeman/1578-fix` (https://github.com/celo-org/celo-blockchain/tree/oneeman/1578-fix) which contains a direct fix for the bug that caused this, with the fix added on top of the commit that you were using when you ran into this.  If you like, you can confirm that it does fix the issue by switching to that branch and syncing to alfajores.  In the meantime, I will create a PR to fix the issue on master.

What the bug was:

https://github.com/celo-org/celo-blockchain/blob/ab78658ae3a738d91d877f289e37fead2a0f1cf7/core/vm/vmcontext/context.go#L27-L39 was not setting the `Header` property as it should.  As a result, when this `vm.Context` which is part of `evm` was used for the gas price minimum call (https://github.com/celo-org/celo-blockchain/blob/ab78658ae3a738d91d877f289e37fead2a0f1cf7/core/state_transition.go#L194), the call was made with header `nil` even though `state` wasn't `nil`.  As a result, https://github.com/celo-org/celo-blockchain/blob/ab78658ae3a738d91d877f289e37fead2a0f1cf7/core/vm/vmcontext/runner.go#L32-L39 would use the current chain's head for the query instead of the state that was passed in.  The block in question included an oracles update followed by a transaction that uses cUSD for the fees.  Because of this bug, the resulting GPM in cUSD was wrong (since it used the cUSD exchange rate from the chain head rather than from the middle of the block, after the oracle update), leading to an incorrect split in the fees between the governance fund and the miner, which is what causes the block to be rejected.

**Comment by oneeman** [comment]
This bad block error has been fixed on master.  It was caused by the `Header` field on `vm.Context` not being set, but with a later refactoring that field was no longer used for that, which fixed it.  The `Header` field not being set has been remedied in #1612 by removing that field.  We also found a bad block on Alfajores block 536, which has now been fixed by #1611.

Thanks again for reporting this!
